### PR TITLE
Replace Crowd with SSO API (Authorisation Service)

### PR DESCRIPTION
**Pull request by maggarwal13**
Replacing direct calls to the underlying Crowd API allows us to switch SSO providers in the future

### CODE REVIEW

File: whois-client/src/main/java/net/ripe/db/whois/common/sso/CrowdClientException.java [hunk 1]
```diff
@@ -9,40 +9,19 @@
 
 public class CrowdClientException extends RuntimeException {
```
Comment (eshryane) [inline whois-client/src/main/java/net/ripe/db/whois/common/sso/CrowdClientException.java:10]: Rename to AuthServiceClientException

File: whois-api/src/test/java/net/ripe/db/whois/api/rest/InternalUpdatePerformerTest.java [hunk 2]
```diff
@@ -186,7 +188,7 @@ public void setSsoSessionToContext_successful_sso_translation() {
 
     @Test
     public void setSsoSessionToContext_exception_is_logged() {
-        when(ssoTokenTranslatorMock.translateSsoToken("test-token")).thenThrow(new CrowdClientException("exception"));
+        when(ssoTokenTranslatorMock.translateSsoToken("test-token")).thenThrow(new CrowdClientException(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode(), "exception"));
```
Comment (eshryane) [inline whois-api/src/test/java/net/ripe/db/whois/api/rest/InternalUpdatePerformerTest.java:191]: Use `HttpStatus.INTERNAL_SERVER_ERROR_500`

File: whois-api/src/test/java/net/ripe/db/whois/api/rest/SyncUpdatesServiceTest.java [hunk 3]
```diff
@@ -31,6 +31,7 @@
 import java.net.HttpURLConnection;
 import java.util.Collections;
 
+import static javax.ws.rs.core.Response.Status.UNAUTHORIZED;
```
Comment (eshryane) [inline whois-api/src/test/java/net/ripe/db/whois/api/rest/SyncUpdatesServiceTest.java:34]: Use `HttpStatus` not `javax.ws.rs.core.Response.Status` for consistency

File: whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+import javax.ws.rs.WebApplicationException;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.core.MediaType;
+import javax.ws.rs.core.Response;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static javax.ws.rs.core.Response.Status.INTERNAL_SERVER_ERROR;
```
Comment (eshryane) [inline whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java:30]: Use `HttpStatus` for consistency

File: whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+import static javax.ws.rs.core.Response.Status.UNAUTHORIZED;
+
+@Component
+public class AuthServiceClient {
+
+    public static final String TOKEN_KEY = "crowd.token_key";
+    private static final Logger LOGGER = LoggerFactory.getLogger(AuthServiceClient.class);
+
+    private static final String VALIDATE_PATH = "/authorisation-service/v2/authresource/validate";
```
Comment (eshryane) [inline whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java:39]: Should the base URL have `/authorisation-service/` in it? Then these relative paths are within the authorisation service.

File: whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java [hunk 6]
```diff
@@ -0,0 +1,197 @@
...
+                .register(jsonProvider))
+                .property(ClientProperties.CONNECT_TIMEOUT, CLIENT_CONNECT_TIMEOUT)
+                .property(ClientProperties.READ_TIMEOUT, CLIENT_READ_TIMEOUT)
+                .build();
+    }
+
+    @Nullable
+    public ValidateTokenResponse validateToken(final String crowdToken) {
+        if (StringUtils.isEmpty(crowdToken)) {
+            LOGGER.info("No crowdToken was supplied");
```
Comment (eshryane) [inline whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java:75]: Replace mentions of 'crowd' with 'authorisation service'

File: whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java [hunk 7]
```diff
@@ -0,0 +1,197 @@
...
+                .configure(DeserializationFeature.ACCEPT_SINGLE_VALUE_AS_ARRAY, true);
+        this.client = (ClientBuilder.newBuilder()
+                .register(jsonProvider))
+                .property(ClientProperties.CONNECT_TIMEOUT, CLIENT_CONNECT_TIMEOUT)
+                .property(ClientProperties.READ_TIMEOUT, CLIENT_READ_TIMEOUT)
+                .build();
+    }
+
+    @Nullable
+    public ValidateTokenResponse validateToken(final String crowdToken) {
```
Comment (eshryane) [inline whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java:73]: rename `crowdToken` to `token`, don't mention Crowd.

File: whois-client/src/main/java/net/ripe/db/whois/common/sso/CrowdClientException.java [hunk 8]
```diff
@@ -9,40 +9,19 @@
...
 
-    private static final Logger LOGGER = LoggerFactory.getLogger(CrowdClientException.class);
+    private final String message;
+    private final int code;
 
-    public CrowdClientException(final String message) {
-        super(message);
+    public CrowdClientException(int code, String message) {
```
Comment (eshryane) [inline whois-client/src/main/java/net/ripe/db/whois/common/sso/CrowdClientException.java]: Use final where possible

File: whois-update/src/main/java/net/ripe/db/whois/update/sso/SsoTranslator.java [hunk 9]
```diff
@@ -54,9 +58,12 @@ public RpslAttribute translate(final String authType, final String authToken, fi
                 if (authType.equals("SSO")) {
                     if (!updateContext.hasSsoTranslationResult(authToken)) {
                         try {
-                            final String uuid = crowdClient.getUuid(authToken);
+                            LOGGER.info("getting details for {}", authToken);
```
Comment (eshryane) [inline whois-update/src/main/java/net/ripe/db/whois/update/sso/SsoTranslator.java]: Log at debug?

Comment (maggarwal13) [reply whois-update/src/main/java/net/ripe/db/whois/update/sso/SsoTranslator.java]: I added this for testing, it has been removed.

File: whois-client/src/main/java/net/ripe/db/whois/common/sso/SsoTokenTranslator.java [hunk 10]
```diff
@@ -6,16 +6,16 @@
...
-        this.crowdClient = crowdClient;
+    public SsoTokenTranslator(final AuthServiceClient authServiceClient) {
+        this.authServiceClient = authServiceClient;
     }
 
-    public UserSession translateSsoToken(final String ssoToken) throws CrowdClientException {
-        final UserSession userSession = crowdClient.getUserSession(ssoToken);
-        userSession.setUuid(crowdClient.getUuid(userSession.getUsername()));
+    public UserSession translateSsoToken(final String ssoToken) {
+        final UserSession userSession = authServiceClient.getUserSession(ssoToken);
```
[the remark this answers is not in the document]
Comment (maggarwal13) [reply whois-client/src/main/java/net/ripe/db/whois/common/sso/SsoTokenTranslator.java:17]: Good point. No. I am not sure why crowd was calling it two times. i will make it a single call

File: whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java (hunk 7 above)
Comment (maggarwal13) [reply whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java:73]: I can rename but the cookie that we get from client still has crowd.token_key, thats why I dint rename this

Comment (eshryane) [reply whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java:73]: We must leave the cookie name as-is, but I think we should remove 'crowd' where we can, and I think 'token' still makes sense as a method argument.

File: whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java (hunk 5 above)
Comment (eshryane) [reply whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java:39]: Thanks, it also makes the code more readable, i.e. we only need relative paths then within the API.

File: whois-api/src/test/java/net/ripe/db/whois/api/rest/InternalUpdatePerformerTest.java (hunk 2 above)
Comment (maggarwal13) [reply whois-api/src/test/java/net/ripe/db/whois/api/rest/InternalUpdatePerformerTest.java]: Authclient is part of whois-client module and we do not have org.eclipse.jetty. In this module we use javax.ws.rs.core.Response.Status.INTERNAL_SERVER_ERROR in other classes as well

File: whois-api/src/test/java/net/ripe/db/whois/api/rest/SyncUpdatesServiceTest.java (hunk 3 above)
Comment (maggarwal13) [reply whois-api/src/test/java/net/ripe/db/whois/api/rest/SyncUpdatesServiceTest.java:34]: Authclient is part of whois-client module and we do not have org.eclipse.jetty. In this module we use javax.ws.rs.core.Response.Status.INTERNAL_SERVER_ERROR in other classes as well

Comment (eshryane) [reply whois-api/src/test/java/net/ripe/db/whois/api/rest/SyncUpdatesServiceTest.java:34]: OK, if whois-client is already different then use `Response.Status` (for now, we should refactor this).

File: whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java [hunk 11]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+    public List<String> getOrgsContactsEmails(final Long membershipId) {
+        if (membershipId != null) {
+            try {
+                final MemberContactsResponse response = client.target(restUrl)
+                        .path(ORGANISATION_MEMBERS_PATH)
+                        .path(String.valueOf(membershipId))
+                        .path(CONTACT_PATH)
+                        .queryParam("api-key", apiKey) // TODO remove this when this endpoint supports apikey as a header
```
Comment (eshryane) [inline whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java:173]: Check if we can replace "api-key" with an API key header

Comment (eshryane) [reply whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java:173]: Confirmed (tested) that the "api-key" query parameter can be replaced with the X-API_KEY header

Comment (maggarwal13) [reply whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java:173]: We are using header X-API_KEY only

File: whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+        } catch (ProcessingException e) {
+            LOGGER.debug("Failed to get details for email {} due to {}:{}", username, e.getClass().getName(), e.getMessage());
+            throw new AuthServiceClientException(INTERNAL_SERVER_ERROR.getStatusCode(), "Internal server error");
+        }
+    }
+
+    public ValidateTokenResponse getUserDetails(final String uuid) {
+        if (StringUtils.isEmpty(uuid)) {
+            LOGGER.info("No uuid was supplied");
+            throw new NotAuthorizedException("Invalid uuid.");
```
Comment (eshryane) [inline whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java:136]: Also `AuthServiceClientException` here?

Comment (maggarwal13) [reply whois-client/src/main/java/net/ripe/db/whois/common/sso/AuthServiceClient.java:136]: Yes, we should

**Review by eshryane — APPROVED** [review]

**Comment by maggarwal13** [comment]
waiting for Integration tests to pass

**Pull request merged**